### PR TITLE
Replace std::unordered_map/set with absl containers

### LINKED ISSUE #11825: Replace std::unordered_map with absl::node_hash_map?
In a number of places throughout the test code, and likely also in user's anticipated behavior of Envoy, it seems that the unordered nature of an unordered map has been ignored. The results are compared to constant strings where the resulting iteration is predicted, and this does not conform to stdc++ unordered map's definition.

For examples, two secrets are documented in an expected list of secrets, and compared to the string dump of an unordered map by SecretManagerImplTest.ConfigDumpHandlerStaticSecrets. In the second case (already fixed) we elected to simplify a configuration to a single element, to avoid this predictable, but inverted ordering that occurs in the libc++/libstdc++ vs msvc cpp runtime library implementations of std::unordered_map.

The later example was discussed at some length https://github.com/envoyproxy/envoy/pull/11335#pullrequestreview-421444954 - however the suggestion to use absl::flat_hash_map turned out to be a non-starter because it is non-deterministic by design. A request to add determinism for purposes such as Envoy seeks turn out to be roundly rejected by the absl authors, see the feedback of the feature request https://github.com/abseil/abseil-cpp/issues/339 for their philosophy.

As a general solution to this and similar observations of small sets for testing that are implemented in Envoy by an unordered_map today, it still seems sensible to migrate to absl::flat_hash_map for the very reason given by the abesil authors. This will obviously break tests which will need to be modified, and possibly could break some user expectations based on expected sequencing. Do we have any thoughts on the Envoy project's approach to this set of issues?

Tagging @mattklein123 and @sunjayBhatia for feedback.

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Commit Message:
Replace std::unordered_map/set with absl containers

- Replace with absl::node_hash_map/set
  - Primarily for performance optimizations and to root out any
     assumptions made about iteration order in tests or otherwise (the
     replacement absl containers have a non-deterministic iteration
     order
  - absl::node_hash_map/set should be drop-in replacements for
     std::unordered_map/set
- Note that a future refactor should reevaluate and move to
   absl::flat_hash_map/set where possible for memory optimizations
- Add format check to disallow future usage of std::unordered_map/set
- Small changes made where absl containers required it or tests needed
   to be modified for correctness

Additional Description:
- There may be an issue we should open with abseil about `emplace` and `try_emplace` when attempting to do in-place construction
  - When a constructor throws an exception, as far as I can tell the c++ language standard says the container should not be affected, however this does not seem to be the case for the absl containers so their guarantees are not the same (though they may be intended to have the same guarantees)
  - //test/server:overload_manager_impl_test demonstrates this
  - see https://github.com/abseil/abseil-cpp/issues/388

Risk Level: Low, absl::node_hash_map/set should be drop-in replacements for std::unordered_map/set though this may shake loose more assumptions in tests over time we weren't able to catch locally
Testing: Small changes to unit tests, repeatedly run on Windows and Linux/clang
Docs Changes: N/A
Release Notes: N/A
Fixes #11825

### CODE REVIEW

**Comment by wrowe** [comment]
Included in 11825 as well, but this should be ready to approve and merge, just waiting on retest of an unrelated asan flake.

**Comment by wrowe** [comment]
> Included in 11825 as well, but this should be ready to approve and merge, just waiting on retest of an unrelated asan flake.

Correction, I transposed the test PR ID's. It is 11825 which should be ready to approve, this one has a bit more WIP.

**Comment by antoniovicente** [comment]
> Notes:
> 
> * Depends on #11875, temporarily cherry picks commit from that PR to get past format check
> * There may be an issue we should open with abseil about `emplace` and `try_emplace` when attempting to do in-place construction

cc: @ahedberg 
Did you file an issue at https://github.com/abseil/abseil-cpp/issues about the emplace exception issue?

>   * When a constructor throws an exception, as far as I can tell the c++ language standard says the container should not be affected, however this does not seem to be the case for the absl containers so their guarantees are not the same (though they may be intended to have the same guarantees)
>   * //test/server:overload_manager_impl_test demonstrates this
> 
> Commit Message:
> Replace std::unordered_map/set with absl containers
> 
> * Replace with absl::node_hash_map/set
>   
>   * Primarily for performance optimizations and to root out any
>     assumptions made about iteration order in tests or otherwise (the
>     replacement absl containers have a non-deterministic iteration
>     order
>   * absl::node_hash_map/set should be drop-in replacements for
>     std::unordered_map/set
> * Note that a future refactor should reevaluate and move to
>   absl::flat_hash_map/set where possible for memory optimizations
> * Add format check to disallow future usage of std::unordered_map/set
> * Small changes made where absl containers required it or tests needed
>   to be modified for correctness
> 
> Additional Description:
> Risk Level: Low, absl::node_hash_map/set should be drop-in replacements for std::unordered_map/set though this may shake loose more assumptions in tests over time we weren't able to catch locally
> Testing: Small changes to unit tests, repeatedly run on Windows and Linux/clang
> Docs Changes: N/A
> Release Notes: N/A
> Fixes #11825

**Comment by ahedberg** [comment]
> > Notes:
> > 
> > * Depends on #11875, temporarily cherry picks commit from that PR to get past format check
> > * There may be an issue we should open with abseil about `emplace` and `try_emplace` when attempting to do in-place construction
> 
> cc: @ahedberg
> Did you file an issue at https://github.com/abseil/abseil-cpp/issues about the emplace exception issue?
https://github.com/abseil/abseil-cpp/issues/388 may be relevant.

> 
> > * When a constructor throws an exception, as far as I can tell the c++ language standard says the container should not be affected, however this does not seem to be the case for the absl containers so their guarantees are not the same (though they may be intended to have the same guarantees)
> > * //test/server:overload_manager_impl_test demonstrates this
> > 
> > Commit Message:
> > Replace std::unordered_map/set with absl containers
> > 
> > * Replace with absl::node_hash_map/set
> >   
> >   * Primarily for performance optimizations and to root out any
> >     assumptions made about iteration order in tests or otherwise (the
> >     replacement absl containers have a non-deterministic iteration
> >     order
> >   * absl::node_hash_map/set should be drop-in replacements for
> >     std::unordered_map/set
> > * Note that a future refactor should reevaluate and move to
> >   absl::flat_hash_map/set where possible for memory optimizations
> > * Add format check to disallow future usage of std::unordered_map/set
> > * Small changes made where absl containers required it or tests needed
> >   to be modified for correctness
> > 
> > Additional Description:
> > Risk Level: Low, absl::node_hash_map/set should be drop-in replacements for std::unordered_map/set though this may shake loose more assumptions in tests over time we weren't able to catch locally
> > Testing: Small changes to unit tests, repeatedly run on Windows and Linux/clang
> > Docs Changes: N/A
> > Release Notes: N/A
> > Fixes #11825

File: source/common/upstream/subset_lb.cc [hunk 1]
```diff
@@ -602,11 +603,15 @@ void SubsetLoadBalancer::purgeEmptySubsets(LbSubsetMap& subsets) {
         stats_.lb_subsets_removed_.inc();
       }
 
-      it = subset_it->second.erase(it);
+      auto next_it = std::next(it);
+      subset_it->second.erase(it);
+      it = next_it;
```
Comment (antoniovicente) [inline source/common/upstream/subset_lb.cc:608]: It may be worth doing a separate PR for non-trivial transformations like this one that require use of std::next to work around the erase return value issue.

Comment (sunjayBhatia) [reply source/common/upstream/subset_lb.cc:608]: There is an erase that takes a range like the stdlib equivalents and returns an iterator which I considered, though it seems worse: https://github.com/abseil/abseil-cpp/blob/38db52adb2eabc0969195b33b30763e0a1285ef9/absl/container/node_hash_map.h#L221
- we would have to use `std::next` to get the end of the range anyways
- the erase method would still do some iteration/extra comparisons even if we're moving the iterator one position in practice
- it seems absl moved to this signature to guarantee an O(1) erase: https://abseil.io/docs/cpp/guides/container#hash-tables `Guarantees an O(1) erase method by returning void instead of an iterator.`

We can pull these related changes into a separate PR

File: include/envoy/grpc/status.h [hunk 2]
```diff
@@ -9,7 +9,7 @@ class Status {
 public:
   using GrpcStatus = int64_t;
 
-  // If this enum is changed, then the std::unordered_map in Envoy::Grpc::Utility::nameToGrpcStatus
+  // If this enum is changed, then the absl::node_hash_map in Envoy::Grpc::Utility::nameToGrpcStatus
```
[the remark this answers is not in the document]
Comment (sunjayBhatia) [reply include/envoy/grpc/status.h]: Removed

File: source/common/common/utility.cc [hunk 3]
```diff
@@ -101,7 +102,9 @@ std::string DateFormatter::fromTime(const SystemTime& time) const {
     // Remove all the expired cached items.
     for (auto it = cached_time.formatted.cbegin(); it != cached_time.formatted.cend();) {
       if (it->second.epoch_time_seconds != epoch_time_seconds) {
-        it = cached_time.formatted.erase(it);
+        auto next_it = std::next(it);
+        cached_time.formatted.erase(it);
+        it = next_it;
       } else {
         it++;
```
Comment (antoniovicente) [inline source/common/common/utility.cc:109]: nit: Should be ++it;

File: source/common/upstream/subset_lb.cc (hunk 1 above)
Comment (antoniovicente) [reply source/common/upstream/subset_lb.cc:608]: After thinking about it some more, I think that the change you're making is equivalent to the original, idiomatic and, a trivial enough refactor.  Keeping it in this PR is fine.

File: source/common/common/utility.cc [hunk 4]
```diff
@@ -101,7 +102,9 @@ std::string DateFormatter::fromTime(const SystemTime& time) const {
     // Remove all the expired cached items.
     for (auto it = cached_time.formatted.cbegin(); it != cached_time.formatted.cend();) {
       if (it->second.epoch_time_seconds != epoch_time_seconds) {
-        it = cached_time.formatted.erase(it);
+        auto next_it = std::next(it);
+        cached_time.formatted.erase(it);
+        it = next_it;
```
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/common/utility.cc:107]: Nevermind, the std::next version is fine.

File: include/envoy/upstream/cluster_manager.h [hunk 5]
```diff
@@ -122,15 +124,15 @@ class ClusterManager {
...
+  using ClusterInfoMap = absl::node_hash_map<std::string, std::reference_wrapper<const Cluster>>;
 
   /**
    * @return ClusterInfoMap all current clusters. These are the primary (not thread local)
    * clusters which should only be used for stats/admin.
    */
   virtual ClusterInfoMap clusters() PURE;
 
-  using ClusterSet = std::unordered_set<std::string>;
+  using ClusterSet = absl::node_hash_set<std::string>;
```
[the remark this answers is not in the document]
Comment (sunjayBhatia) [reply include/envoy/upstream/cluster_manager.h]: pushed a change to move this to flat_hash_set, will see how this performs

**Comment by antoniovicente** [comment]
Please merge and remove draft status when you have a chance.

**Comment by sunjayBhatia** [comment]
fixing conflicts now

File: source/server/overload_manager_impl.cc [hunk 6]
```diff
@@ -143,8 +142,8 @@ OverloadManagerImpl::OverloadManagerImpl(Event::Dispatcher& dispatcher, Stats::S
   for (const auto& action : config.actions()) {
     const auto& name = action.name();
     ENVOY_LOG(debug, "Adding overload action {}", name);
-    auto result = actions_.emplace(std::piecewise_construct, std::forward_as_tuple(name),
-                                   std::forward_as_tuple(action, stats_scope));
+    // Cannot use in place construction as OverloadAction constructor may throw
+    auto result = actions_.try_emplace(name, OverloadAction(action, stats_scope));
```
Comment (antoniovicente) [inline source/server/overload_manager_impl.cc:146]: Are there other places where the emplace and exception interactions come up?  Is there an upstream absl issue tracking making the emplace methods exception safe?  The status quo seems risky.

File: tools/code_format/check_format_test_helper.py [hunk 7]
```diff
@@ -232,6 +232,12 @@ def runChecks():
       "Registry::InjectFactory instead.")
   errors += checkUnfixableError("strerror.cc",
                                 "Don't use strerror; use Envoy::errorDetails instead")
+  errors += checkUnfixableError(
+      "std_unordered_map.cc", "Don't use std::unordered_map; use absl::flat_hash_map instead " +
```
Comment (antoniovicente) [inline tools/code_format/check_format_test_helper.py:236]: We seem to be preferring the node_hash_map, not flat_hash_map.

File: tools/code_format/check_format_test_helper.py [hunk 8]
```diff
@@ -232,6 +232,12 @@ def runChecks():
...
+      "or absl::node_hash_map if pointer stability of keys/values is required")
+  errors += checkUnfixableError(
+      "std_unordered_set.cc", "Don't use std::unordered_set; use absl::flat_hash_set instead " +
+      "or absl::node_hash_set if pointer stability of keys/values is required")
```
Comment (antoniovicente) [inline tools/code_format/check_format_test_helper.py:240]: Is it too ambitious to enable these lint checks in the same PR as the large scale change above?  It may be good to do some smaller PRs which can be reviewed quickly to adjust the code base in preparation for these linter changes.

File: source/server/overload_manager_impl.cc (hunk 6 above)
Comment (sunjayBhatia) [reply source/server/overload_manager_impl.cc:146]: Doing a quick grep, the only place I see the piecewise construction happening with `emplace` is in this file and since we just moved to C++17, we don't seem to have any usage of `try_emplace` yet

Is this absl issue sufficient? https://github.com/abseil/abseil-cpp/issues/388 we can report another

File: tools/code_format/check_format_test_helper.py (hunk 8 above)
Comment (sunjayBhatia) [reply tools/code_format/check_format_test_helper.py:240]: Risky in the sense that existing PRs may be introducing new instances of stdlib unordered containers and depending on merge order master will be broken?

File: tools/code_format/check_format_test_helper.py (hunk 7 above)
Comment (sunjayBhatia) [reply tools/code_format/check_format_test_helper.py:236]: Ah, we put `flat_hash_map` as the first choice since it is recommended for new code and `node_hash_map` is the stabler drop-in replacement for `std::unordered_map`, we can flip the order or remove one option if you like

Comment (wrowe) [reply tools/code_format/check_format_test_helper.py:236]: I prefer we emphasize flat_hash_map. The node_hash_map makes certain additional guarantees that shouldn't be needed in new code.

File: include/envoy/upstream/cluster_manager.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (sunjayBhatia) [reply include/envoy/upstream/cluster_manager.h]:
```

Stack trace:
  0x90e5a9: Envoy::(anonymous namespace)::ClusterMemoryTestRunner_MemoryLargeClusterSizeWithRealSymbolTable_Test::TestBody()
  0x174a378: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x174a2a5: testing::Test::Run()
  0x174b0f0: testing::TestInfo::Run()
... Google Test internal frames ...

[  FAILED  ] IpVersions/ClusterMemoryTestRunner.MemoryLargeClusterSizeWithRealSymbolTable/IPv4, where GetParam() = 4-byte object <00-00 00-00> (694 ms)
```

**Comment by sunjayBhatia** [comment]
cc @davinci26 @nigriMSFT

File: include/envoy/upstream/cluster_manager.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (antoniovicente) [reply include/envoy/upstream/cluster_manager.h]: seems better

File: source/server/overload_manager_impl.cc (hunk 6 above)
Comment (antoniovicente) [reply source/server/overload_manager_impl.cc:146]: Reference the issue in the code.  How did you detect this exception issue?  Should we avoid use of try_emplace until the upstream issue is fixed?

File: tools/code_format/check_format_test_helper.py (hunk 8 above)
Comment (antoniovicente) [reply tools/code_format/check_format_test_helper.py:240]: You need to be merging constantly to get this large change to land.  Adding this lint checks in a smaller, faster to review PR will likely reduce the difficulty of getting the change merged in.

File: source/server/overload_manager_impl.cc (hunk 6 above)
Comment (sunjayBhatia) [reply source/server/overload_manager_impl.cc:146]: adding a note about that issue and the symptoms

we detected this in a unit test failure: https://github.com/envoyproxy/envoy/blob/142b27533ba110afc13651bd75d05b4e25fa1c0d/test/server/overload_manager_impl_test.cc#L359-L382, when the exception is thrown and memory is freed we got an invalid free

With proper test coverage of any constructor that may throw an exception, I would guess this issue would get caught with a test failure, any in-place construction of a type whose constructor does not throw is safe

Comment (antoniovicente) [reply source/server/overload_manager_impl.cc:150]: I don't don't know if we have full test coverage for every constructor that can throw an exception which is used as a map value.  I think the main place where this could come up is the control plane doing config validation in constructors while loading config.

File: tools/code_format/check_format_test_helper.py (hunk 8 above)
Comment (wrowe) [reply tools/code_format/check_format_test_helper.py:240]: That would prevent this commit from potentially catching newly introduced violations of this rule at the time we adopt the change. Better that we do merge, and correct the newly introduced exceptions

File: tools/code_format/check_format_test_helper.py (hunk 7 above)
Comment (antoniovicente) [reply tools/code_format/check_format_test_helper.py:236]: Recommending flat_hash_map seems fine.  It may be worth pointing to docs in cases where flat_hash_map doesn't work.

Do we intend to do a second pass to change node_hash_map use to flat_hash_map where doing so is safe?

File: tools/code_format/check_format_test_helper.py (hunk 8 above)
Comment (antoniovicente) [reply tools/code_format/check_format_test_helper.py:240]: I guess the rate at which we add new uses of ordered maps is not high enough to be a problem, so it may be fine to start enforcing immediately. I think the open comment that remains is the one about exception safety.

File: tools/code_format/check_format_test_helper.py (hunk 7 above)
Comment (wrowe) [reply tools/code_format/check_format_test_helper.py:236]: We certainly should, but the windows port is unblocked by merging this current iteration. It sounds like a perfect Help Wanted issue for a contributor with the round tuits to spend. We have a couple months ahead of the more convoluted fixes to work through before looping back on good tickets like this.

Comment (wrowe) [reply tools/code_format/check_format_test_helper.py:236]: (Actually, a Help Wanted and Tech Debt tag on that follow up issue)

File: tools/code_format/check_format_test_helper.py (hunk 8 above)
Comment (sunjayBhatia) [reply tools/code_format/check_format_test_helper.py:240]: ok sweet, gonna resolve this comment and we can focus on the exception safety solution

**Comment by sunjayBhatia** [comment]
@antoniovicente do you happen to know a hint to fix these clang-tidy errors in `tools/clang_tools/api_booster`? it seems like the compilation db etc. is already generated for those targets, not sure how else to get it to see `gtest.h` etc.

**Comment by antoniovicente** [comment]
> @antoniovicente do you happen to know a hint to fix these clang-tidy errors in `tools/clang_tools/api_booster`? it seems like the compilation db etc. is already generated for those targets, not sure how else to get it to see `gtest.h` etc.

cc @lizan 

I don't.  The clang errors seem pre-existing and not actionable.

File: include/envoy/upstream/cluster_manager.h [hunk 9]
```diff
@@ -123,15 +125,15 @@ class ClusterManager {
...
+  using ClusterInfoMap = absl::node_hash_map<std::string, std::reference_wrapper<const Cluster>>;
 
   /**
    * @return ClusterInfoMap all current clusters. These are the primary (not thread local)
    * clusters which should only be used for stats/admin.
    */
   virtual ClusterInfoMap clusters() PURE;
 
-  using ClusterSet = std::unordered_set<std::string>;
+  using ClusterSet = absl::flat_hash_set<std::string>;
```
Comment (lizan) [inline include/envoy/upstream/cluster_manager.h:136]: are we sure this is safe?

**Comment by lizan** [comment]
> > @antoniovicente do you happen to know a hint to fix these clang-tidy errors in `tools/clang_tools/api_booster`? it seems like the compilation db etc. is already generated for those targets, not sure how else to get it to see `gtest.h` etc.
> 
> cc @lizan
> 
> I don't. The clang errors seem pre-existing and not actionable.

In large replace PR like this clang-tidy might pick pre-existing errors, feel free to ignore this.

File: include/envoy/upstream/cluster_manager.h (hunk 9 above)
Comment (wrowe) [reply include/envoy/upstream/cluster_manager.h:136]: Could you define safe? Yes we don't see regressions. The contract of unordered_set is that the order is indeterminate. We've already shown up that the hash values constructed by msvc Cary from those of gcc or clang. We know from the test changes that many coders take an easy way out with the treatment of unordered data sets, and the abesil team has gone to great lengths by adding entroy to prove up such foolish assumptions. Yes, the patch is ready to accept and merge. Where exceptions can be thrown by elemental constructors, we have to deal with those, but we think we've found them all. That's a legit bug that abesil should address and already has an open ticket. As far as predictability of behavior of envoy based on unordered associations in end user configs... It seems time to bite the bullet and merge this, and then introduce still-breaking changes by introducing contracted behavioral ordering based on sorted collections as needed. 1.15 just dropped, so the time to begin evaluating this for 1.16 on a wide scale was yesterday.

Comment (lizan) [reply include/envoy/upstream/cluster_manager.h:136]: I'm not asking about the whole of this PR, was just asking why this is the only place that you chose absl::flat_hash_set over absl::node_hash_set. Did you checked all usage of this type that doesn't rely on pointer stability?

Comment (sunjayBhatia) [reply include/envoy/upstream/cluster_manager.h:136]: We moved to `flat_hash_set` to try to reduce memory usage that the stats integration tests showed was increased because of this change (see this thread: https://github.com/envoyproxy/envoy/pull/11879#discussion_r450429690)

I believe I checked if this set was used w/o a need for pointer stability but ill look again, we haven't seen any test coverage show anything i believe yet

Comment (lizan) [reply include/envoy/upstream/cluster_manager.h:136]: Thanks for clarification, SGTM.
